position/position.gno

```diff
@@ -96,44 +92,199 @@ func mint(params MintParams) (uint64, bigint, bigint, bigint) {
 	return tokenId, liquidity, amount0, amount1
 }
 
-func CollectFee(tokenId uint64) (uint64, bigint, bigint, string) { // tokenId, tokensOwed0, tokensOwed1, poolPath
+func IncreaseLiquidity(
+	tokenId uint64,
+	amount0Desired bigint,
+	amount1Desired bigint,
+	amount0Min bigint,
+	amount1Min bigint,
+	deadline bigint,
+) (uint64, bigint, bigint, bigint) { // tokenId, liquidity, amount0, amount1
+	increaseLiquidityParams := IncreaseLiquidityParams{
+		tokenId:        tokenId,
+		amount0Desired: amount0Desired,
+		amount1Desired: amount1Desired,
+		amount0Min:     amount0Min,
+		amount1Min:     amount1Min,
+		deadline:       deadline,
+	}
+
+	return increaseLiquidity(increaseLiquidityParams)
+}
+
+func increaseLiquidity(params IncreaseLiquidityParams) (uint64, bigint, bigint, bigint) {
 	// verify tokenId exists
-	require(exists(tokenId), ufmt.Sprintf("[POSITION] position.gno__CollectFee() || tokenId(%d) doesn't exist", tokenId))
+	require(exists(params.tokenId), ufmt.Sprintf("[POSITION] position.gno__increaseLiquidity() || tokenId(%d) doesn't exist", params.tokenId))
 
-	// verify owner or approved
-	isAuthorizedForToken(tokenId)
+	// MUST BE OWNER TO DECREASE LIQUIDITY
+	// can not be approved address > staked position can be modified
+	owner := gnft.OwnerOf(tid(params.tokenId))
+	require(owner == GetOrigCaller(), ufmt.Sprintf("[POSITION] position.gno__decreaseLiquidity() || only owner can decrease liquidity__owner(%s) == GetOrigCaller(%s)", owner, GetOrigCaller()))
 
-	position := positions[tokenId]
-	tokensOwed0, tokensOwed1 := position.tokensOwed0, position.tokensOwed1
+	checkDeadline(params.deadline)
 
-	pToken0, pToken1, pFee := poolKeyDivide(position.poolKey)
-	pl.Burn(pToken0, pToken1, pFee, position.tickLower, position.tickUpper, 0) // burn '0' liquidity to collect fee
+	position := positions[params.tokenId]
+	liquidity, amount0, amount1 := addLiquidity(
+		AddLiquidityParams{
+			poolKey:        position.poolKey,
+			recipient:      GetOrigPkgAddr(), // MUST BE POSITION
+			tickLower:      position.tickLower,
+			tickUpper:      position.tickUpper,
+			amount0Desired: params.amount0Desired,
+			amount1Desired: params.amount1Desired,
+			amount0Min:     params.amount0Min,
+			amount1Min:     params.amount1Min,
+		},
+	)
+
+	pool := pl.GetPoolFromPoolPath(position.poolKey)
 
 	positionKey := positionKeyCompute(GetOrigPkgAddr(), position.tickLower, position.tickUpper)
+
+	feeGrowthInside0LastX128, feeGrowthInside1LastX128 := pool.PoolGetPositionFeeGrowthInside0LastX128(positionKey), pool.PoolGetPositionFeeGrowthInside1LastX128(positionKey)
+
+	position.tokensOwed0 += (feeGrowthInside0LastX128 - position.feeGrowthInside0LastX128) * position.liquidity / consts.Q128
+	position.tokensOwed1 += (feeGrowthInside1LastX128 - position.feeGrowthInside1LastX128) * position.liquidity / consts.Q128
```

Comment:
It would be good to store the calculation result of `liquidity / Q128`

```diff
- position.tokensOwed0 += (feeGrowthInside0LastX128 - position.feeGrowthInside0LastX128) * position.liquidity / consts.Q128
- position.tokensOwed1 += (feeGrowthInside1LastX128 - position.feeGrowthInside1LastX128) * position.liquidity / consts.Q128
+ liquidityQuotient := position.liquidity / consts.Q128
+ position.tokensOwed0 += (feeGrowthInside0LastX128 - position.feeGrowthInside0LastX128) * liquidityQuotient
+ position.tokensOwed1 += (feeGrowthInside1LastX128 - position.feeGrowthInside1LastX128) * liquidityQuotient